pkg/blobstore/cassandra/cassandra_blob_access.go

```diff
@@ -0,0 +1,673 @@
+	}, "-")
+
+	estimatedSegmentCount := getSegmentCount(a.segmentSize, digest.GetSizeBytes())
+
+	now := time.Now()
+	var segmentCount int
+
+	if a.tables.content.isSegmentPresent(ctx, blobID, estimatedSegmentCount) {
+		segmentCount = estimatedSegmentCount
+	} else {
+		instanceName := a.getInstanceName(digest)
+		if err := a.tables.orphanedContent.insert(ctx, digest, blobID, instanceName, estimatedSegmentCount, now); err != nil {
+			b.Discard()
+			return err
+		}
+		defer a.tables.orphanedContent.delete(ctx, blobID, instanceName)
+
+		actualCount, err := a.writeContentData(ctx, digest, b, blobID)
+		if err != nil {
+			return err
+		}
+		segmentCount = actualCount
+	}
+
+	return a.tables.metadata.update(ctx, digest, blobID, a.getInstanceName(digest), segmentCount, now, a.segmentSize)
+}
+
+func getSegmentCount(segmentSize int32, sizeInBytes int64) int {
+	segmentSize64 := int64(segmentSize)
+
+	return int((sizeInBytes + segmentSize64 - 1) / segmentSize64)
+}
+
+func (a *cassandraBlobAccess) writeContentData(ctx context.Context, digest bbdigest.Digest, b buffer.Buffer, blobID string) (int, error) {
+	segmentCount := 0
+
+	g, gCtx := errgroup.WithContext(ctx)
+	g.SetLimit(simultaneousWorkers + 1)
+
+	g.Go(func() error {
+		reader := b.ToReader()
+		defer reader.Close()
+
+		for {
+			// Read either up to EOF or until we have `segmentSize` bytes
+			buf := new(bytes.Buffer)
+			_, err := io.CopyN(buf, reader, int64(a.segmentSize))
+			isEOF := err == io.EOF
+
+			if err != nil && err != io.EOF {
+				logErrorIfNotCancelledContext(err, "Put: Failed to read next segment to write from client. %d of %s (digest function: %s, size: %d)", segmentCount, digest.GetHashString(), digest.GetDigestFunction().GetEnumValue().String(), digest.GetSizeBytes())
+				return err
+			}
+
+			count := segmentCount
+			g.Go(func() error {
+				return a.tables.content.insertSegment(gCtx, blobID, buf.Bytes(), count, digest.GetHashString())
+			})
+
+			segmentCount++
+
+			if isEOF {
+				break
+			}
+		}
+
+		// TODO: Ensure hashes match at this point
+		// TODO: Delete from the blobs table if hashes do not match
+
+		return nil
+	})
+
+	err := g.Wait()
+	if err != nil {
+		logErrorIfNotCancelledContext(err, "Put: Unable to store %s", digest.String())
+		return 0, err
+	}
+	return segmentCount, nil
+}
+
+func (a *cassandraBlobAccess) FindMissing(ctx context.Context, digests bbdigest.Set) (bbdigest.Set, error) {
+	// The obvious thing to do here is to create a `SELECT ... IN (?, ?)` for each digest, but doing so
+	// prevents token-aware routing of queries. Instead, we issue a single `SELECT` per digest and
+	// aggregate the values that come back.
+
+	builder := bbdigest.NewSetBuilder()
+	var mu sync.Mutex
+
+	g, errCtx := errgroup.WithContext(ctx)
+	// Try not to be a terrible, terrible citizen
+	g.SetLimit(simultaneousWorkers)
+	for _, digest := range digests.Items() {
+		g.Go(func() error {
+			return a.isMissing(errCtx, digest, &builder, &mu)
+		})
+	}
```

Comment:
A downside of using .SetLimit() in combination with a loop like this is that there is no cancelation at all during loop iteration. If a single `.isMissing()` fails, this code will fail to break immediately, meaning all remaining entries MUST be checked, even if we know they will fail.

Please consider getting rid of the `.SetLimit()` call and using a separate semaphore, like we do in most other places of the Buildbarn codebase.

Reply:
How's that?